### DISCUSSION TITLE
Isolate print job generator in separate package to allow easy re-use?

### DISCUSSION THREAD

**Discussion by holtkamp** [star-micronics/star-cloudprnt-for-woocommerce#11] [Ideas]
We are currently investigating the use of the [Star CloudPRNT Protocol](https://www.star-m.jp/products/s_print/CloudPRNTSDK/Documentation/en/developerguide/introduction.html) to have remote Star printers print orders that are placed using our in-house developed e-commerce platform.

Installing a tool like [`cputil`](https://www.star-m.jp/products/s_print/CloudPRNTSDK/Documentation/en/articles/cputil/cputil.html) can be a challenge on Cloud Application Platforms like [Heroku](https://www.heroku.com). In search of a way to circumvent the use of `cputil` I encountered this plugin for WooCommerce which [seems to be able to generate print jobs directly / without using `cputil`](https://github.com/star-micronics/star-cloudprnt-for-woocommerce/blob/master/development.md#generating-print-jobs---order-handlerphp-and-the-printer-object), right? 

Would it be an idea to isolate the functionality that generates print jobs in a separate package and publish it on [Packagist.org](https://packagist.org) to allow users to use [Composer](https://getcomposer.org) to include it in their projects?

Would this be something that can be considered? Or does it already exist?

I also noticed this thread which also covers this subject a bit: https://wordpress.org/support/topic/add-custom-images-2/

**Comment by Lozrus** [comment]
Hi @holtkamp 
Yes, hosting cputil is easy on your own servers, but not always ideal in other cases, which is one reason that it's not used by the WooCommerce plugin. A pure PHP option would be useful for many and is definitely something that I've considered ((but doesn't exist currently, sorry).

I think some extra work is needed to package it up well, but I'm tidying up the print job generator/document builder APi code a lot in the 2.1 branch of this plugin. It is also being documented as an API that can be used by WordPress/WooCommerce plugin developers to extend the print job. So at least we are making a start.

In the mean-time, I suggest that you pull the `cloudprnt/cloudprnt_builde*` and `cloudprnt/printer_star*` files from the 2.1 branch of this project and use the builder class directly. I think it should be reasonable self contained, other than a few constant values defined elsewhere in the code.

Alternatively, for a super easy (but paid) option, Star also have a live cloudPRNT hosting service, which you can use. You just need to connect your printers to this service, instead of to your own, and then use our[ REST API](https://docs.starprinter.online/) to send print jobs. Printing in this case just requires an HTTP POST, with the correckt API key and the print job as HTML, Star Markup, PNG, Jpeg etc. So it is very easy, and requires no special dependencies or services to be hosted by you. This is currently only available in Europe, Middle East and African sales regions though.

**Comment by holtkamp** [comment]
Hi @Lozrus, first of al: thanks for taking the time to write such a detailed response! Great to see 😉 

> A pure PHP option would be useful for many and is definitely something that I've considered ((but doesn't exist currently, sorry)

No need to be sorry at all, it seems you are already doing a great job on this subject. 

> I think some extra work is needed to package it up well, but I'm tidying up the print job generator/document builder APi code a lot in the 2.1 branch of this plugin. It is also being documented as an API that can be used by WordPress/WooCommerce plugin developers to extend the print job. So at least we are making a start.

Nice, I am willing to help on getting this "native" approach in the air by providing feedback, testing, etc. I know broad support for PHP versions is probably required, but from my experience being "strict" (such as using typed properties, available from PHP `7.4` eases development and prevents bugs, also by using tools like https://phpstan.org

What PHP version are you thinking of as minimum requirement?

> In the mean-time, I suggest that you pull the cloudprnt/cloudprnt_builde* and cloudprnt/printer_star* files from the 2.1 branch of this project and use the builder class directly. I think it should be reasonable self contained, other than a few constant values defined elsewhere in the code.

Good suggestion, will dive into this and use these files to give it a try.

> Alternatively, for a super easy (but paid) option, Star also have a live cloudPRNT hosting service, which you can use. You just need to connect your printers to this service, instead of to your own, and then use our REST API to send print jobs. Printing in this case just requires an HTTP POST, with the correckt API key and the print job as HTML, Star Markup, PNG, Jpeg etc. So it is very easy, and requires no special dependencies or services to be hosted by you. This is currently only available in Europe, Middle East and African sales regions though.

Ah, good idea, I was not aware of this service yet. Maybe it is good to have https://docs.starprinter.online mentioned on a page like https://www.star-m.jp/products/s_print/CloudPRNTSDK/Documentation/en/index.html as well? Or is it still "under the radar" because of the beta status?

To create an account, I read: "Please contact your regional Star Micronics office to request an account.", I will contact the EMEA office and try to see how smooth the onboarding process goes 😉 

Looking forward working on this!

**Reply by Lozrus** [reply]
@holtkamp I hope you don't mind, but I converted this to a discussion topic, so that conversation can continue without leaving an open issue.

I'm very far from a PHP expert (I have never used it before inheriting this project), but since I come from a background of C++/C# and other more strictly typed languages, then I 100% agree with you. The WordPress plugin version may need to retain 7.2 compatibility for a little while longer, but for a new library, I'd be completely happy to set a 7.4 minimum.

Thank you for the offer of help, it may be a few weeks before anything can happen, since my primary commitment now is the starprinter.online spooling service, and I need to get the WordPress/WooCommerce 2.1 update out as a priority. But I'll be sure to post something to this thread as and when there is something to try, and a GitHub repository.

If you have any questions about using the document builder classes from the this repository though, please feel free to ask here, or via the Star EMEA tech support, whichever seems appropriate.

> Or is it still "under the radar" because of the beta status?

Yes, exactly that for now, partially because of the beta status, and partially because it is currently only available in the EMEA region. So we are simply suggesting it to developers where it seems like it may be an appropriate solution for now.

> To create an account, I read: "Please contact your regional Star Micronics office to request an account.", I will contact the EMEA office and try to see how smooth the onboarding process goes

They should be able to get you set-up and running very quickly, but if you have any issues at all then let us know.

**Reply by holtkamp** [reply]
@Lozrus, I did not even know Github supported "discussions" 😉 , good idea!

This weekend I had a first look and combined this repository with work of @onkbear found here: https://github.com/onkbear/laravel-star-cloud-prnt/tree/master/src/Buffers

One of the things I encountered: is there an overview to see what functions are supported for [these MediaTypes](https://www.star-m.jp/products/s_print/CloudPRNTSDK/Documentation/en/developerguide/contentmediatypes/contentmediatypes_overview.html).

For example, what MediaTypes support:
- sounding the buzzer
- full paper cut
- partial paper cut
- etc...

Now this has to be derived from the source code. A table/matrix which describes for each MediaType what functionality is supported would be nice to have 🤓 

Another thing that came to mind: instead of using inheritance (X extends Y) for the buffers / printJobBuilder, it might be good to use [PHP Traits](https://www.php.net/manual/en/language.oop5.traits.php) to "sprinkle" functionality over the various kinds of MediaTypes / Buffers.

**Reply by holtkamp** [reply]
@Lozrus an additional question: what would be the right place to submit issues regarding Start CloudPRNT?

I encountered a situation where the `jobGetUrl` and `jobConfirmationUrl` of JSON Response to a Polling request seem to be ignored so I wanted to submit an issue. 

Maybe the example/documentation on https://www.star-m.jp/products/s_print/CloudPRNTSDK/Documentation/en/developerguide/pollingserver/post_jsonresponse.html is outdated?

**Reply by Lozrus** [reply]
@holtkamp 

For general support on the CloudPRNT protocol, there is no directly linked GitHub repository, so please contact Star technical support directly.

From looking, the documentation is ok, but I think that the example JSON is a bit ahead of itself. The `jobGetUrl` and `jobConfirmationUrl` fields are more recent additions to  the CloudPRNT spec and not yet supported by all devices. If you are using a printer with HIX interface, then recent firmware will support this, but if you are using an mC-Print, then I believe that it is not yet available, but hopefully will be added in a future update.

I'm actually on holiday this week, and can't easily check the current status of mC-Print support for the newer CloudPRNT spec features, but the tech support guys should be able to verify that for you. Once I get back I'll arrange for that documentation to be clarified.

For now though, I would recommend assuming that the printer will connect to the same URL for the POST, GET and DELETE requests.
